src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php

```diff
@@ -0,0 +1,208 @@
+    {
+        // Super admins have access to all shops and that cannot be changed by the user.
+        if ($data['profile'] == $this->superAdminProfileId) {
+            $data['shop_association'] = $this->defaultShopAssociation;
+        }
+
+        /** @var EmployeeId $employeeId */
+        $employeeId = $this->bus->handle(new AddEmployeeCommand(
+            $data['firstname'],
+            $data['lastname'],
+            $data['email'],
+            $data['password'],
+            $data['optin'],
+            $data['default_page'],
+            $data['language'],
+            $data['active'],
+            $data['profile'],
+            isset($data['shop_association']) ? $data['shop_association'] : $this->defaultShopAssociation
+        ));
+
+        return $employeeId->getValue();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($id, array $data)
+    {
+        $command = (new EditEmployeeCommand($id))
+            ->setFirstName(new FirstName($data['firstname']))
+            ->setLastName(new LastName($data['lastname']))
+            ->setEmail(new Email($data['email']))
+            ->setIsSubscribedToNewsletter((bool) $data['optin'])
+            ->setDefaultPageId((int) $data['default_page'])
+            ->setLanguageId((int) $data['language'])
+            ->setActive((bool) $data['active'])
+            ->setProfileId((int) $data['profile'])
+        ;
+
+        if ($this->employeeFormAccessChecker->isRestrictedAccess((int) $id)) {
+            if ($this->shouldChangePassword($data)) {
+                $this->assertPasswordIsSameAsOldPassword(
+                    $data['change_password']['old_password'],
+                    $id
+                );
+
+                $command->setPlainPassword(new Password($data['change_password']['new_password']));
+            }
+        } elseif (isset($data['password'])) {
+            $command->setPlainPassword(new Password($data['password']));
+        }
+
+        if (isset($data['shop_association'])) {
+            $shopAssociation = $data['shop_association'] ?: [];
+            $command->setShopAssociation(
+                array_map(function ($shopId) { return (int) $shopId; }, $shopAssociation)
+            );
+        }
+
+        /** @var EmployeeId $employeeId */
+        $employeeId = $this->bus->handle($command);
+
+        return $employeeId->getValue();
+    }
+
+    /**
+     * Asserts if given password is the same as employee's password.
+     *
+     * @param string $plainPassword
+     * @param int $employeeId
+     *
+     * @throws EmployeeConstraintException
+     */
+    private function assertPasswordIsSameAsOldPassword($plainPassword, $employeeId)
+    {
+        $oldPassword = $this->employeeDataProvider->getHashedPasswordById($employeeId);
+
+        if (!$this->hashing->checkHash($plainPassword, $oldPassword)) {
+            throw new EmployeeConstraintException(
+                'Old password is invalid.',
```

Comment:
Not accurate. It's rather
```suggestion
                'Old and new passwords do not match.',
```